### ISSUE TITLE
ECH and AltSvc

### ISSUE THREAD

**Issue by martinthomson** [MikeBishop/dns-alt-svc#326]
I find the text in the current draft a little too light when it comes to ECH and AltSvc.  I find that I can't work out exactly what to do based on the text that is present.

Each is fine in isolation:

* Section 9.1 makes it clear that when a client supports ECH it makes SVC queries always.  If it *requires* ECH then it will insist on receiving an ECH config.  In either case, if there is an ECH config, not getting ECH results in hard failure.

* Section 8.3 makes it clear that for AltSvc, you just use A/AAAA to find the server.  But that means that you won't have an ECH config.

The only text about the combination is this text in Section 8.3:

> Clients that implement support for ECH MUST perform the HTTPS RR query first, and MUST only make use of Alt-Svc when operating in SVCB-optional mode (see Section 9.1).

So, what name is queried for?  I *think* that this refers to the origin.  That is, the client needs to confirm whether the origin has an ECH config in SVCB.  If the ECH config exists, then the client needs to use it (and so it can't use AltSvc).  If the ECH config doesn't exist, then it is free to use AltSvc.

This is a valid conclusion, but I haven't seen any real discussion on this point (to be fair, I could have missed it, the volume of messages on this draft means that I've not read all of them).

I can see why you would prohibit the use of AltSvc if ECH is enabled for the origin.  AltSvc requires that the SNI indicate the origin in the TLS handshake, the origin has indicated (using SVCB) that it wants that information protected, and no means has been provided to protect that information.  I can see other options:

1. We might allow the origin to indicate that its ECH config applies to alternative services.  That could be a sign that the ECH config at the origin equally applies to all alternatives.  This would be a good outcome in my opinion, but I would need to walk through the implications for distributed deployments.  I think that it's OK because you only get AltSvc after successfully talking to a server and presumably one that has used the ECH config successfully.

2. We might allow the alternative service to offer its own ECH config and so offer that up through SVCB.  This might need some sort of indication from the origin (as part of AltSvc perhaps, as part of SVCB, or as part of its own ECH config) to delegate this.  In this case, the client requests SVCB for the alternative.  I have some reservations about taking this path; I haven't considered all the implications fully and it seems like there might be a lot of them.

Whatever the outcome is here, I'm uncomfortable with the level of treatment of this problem in the draft.  It needs more text and - ideally - more debate.

The current design has the effect of disabling alternative services in their current form if you deploy ECH.  You might argue that SVCB is largely a replacement for AltSvc, but I would disagree on the basis that AltSvc provides more precise and timely control over how requests are distributed.

**Comment by enygren** [comment]
At least for some of this our thinking was that it would be preferable to do an AltSvc-bis that covered these topics.  As we got into them they started getting too deep into details of AltSvc and updating AltSvc in this draft seemed out-of-scope.  We opened #246 to track some AltSvc-related future updates, although that only has one item in it.  (I've started a notes file with potential things to consider for AltSvc-bis).

In particular, it seems like most of the ways to do ECH and AltSvc in combination require a draft that updates AltSvc.   We should make sure to not preclude that here, and perhaps we should be more clear in the text that the Section 8.3 text may be modified by a future AltSvc draft.  I'm not sure what the best way to handle this is?

**Comment by martinthomson** [comment]
Thanks for the context @enygren.  I think that no matter what outcome this has, it needs to be better documented than it currently is.  I don't *like* deferral much as it effectively renders AltSvc useless until it is fixed, or at least means that servers need to choose between AltSvc and ECH in that period.  I might be convinced to accept that if we knew approximately the shape of the thing we wanted and we just split the work out so that we can make progress.

**Comment by ericorth** [comment]
I must have missed this implication in my previous reads through the draft...

What is the logic behind not querying HTTPS when connecting to an AltSvc-provided hostname? If it's some concern around avoiding redirection cycles, I don't think that would be an issue in practice, no more so than a pure AltSvc redirection cycle.  At least in our implementation, once we start attempting a connection to an AltSvc-provided hostname, any new AltSvc is at worst just going to change the hostname used for the next request.

Until this thread, my plan for Chrome was that when we had AltSvc info with a different hostname, we would start an h3 connection attempt against it (Chrome only supports connecting to AltSvc hostnames via h3), query A/AAAA/HTTPS, and follow any HTTPS redirection to h3-supporting endpoints, or if no such endpoints, fallback to h3 connection to the A/AAAA.  Then if all that takes too long (or fails), start a racing fallback to attempt an h3 connection attempt against any h3-supporting endpoints found via HTTPS queries for the request hostname.  Maybe I'm missing something reason this is bad, because this still seems like the best plan to me.

**Comment by bemasc** [comment]
> Whatever the outcome is here, I'm uncomfortable with the level of treatment of this problem in the draft.

OK, let's expand the discussion of that interaction.

> I don't like deferral much as it effectively renders AltSvc useless until it is fixed, or at least means that servers need to choose between AltSvc and ECH in that period.

I would quibble with this.  Servers can safely publish both; it's clients that make the choice to use ECH (and thus lose support for Alt-Svc).  Making ECH secure without impairing Alt-Svc for non-ECH clients was a key design goal.

> What is the logic behind not querying HTTPS when connecting to an AltSvc-provided hostname?

Intuitively, that would seem weird to me.  SVCB is meant to parallel Alt-Svc, providing an alternative endpoint for the origin with relevant connection metadata.  Stacking them raises a lot of questions, like what is the client supposed to do with conflicting metadata?  The Alt-Svc metadata is better targeted, but there some things (like ECH support) that it currently can't express, and divergence is likely to grow over time as new parameters are defined.  However, this does seem like a behavior that could be defined by a future draft.

For background, early versions of this draft solved this by putting an ECHConfig in Alt-Svc as well as SVCB, but that proved controversial so we took it out.  We could revisit this in a revision of Alt-Svc ... or perhaps some experience with SVCB will show that Alt-Svc is not needed when SVCB is in use.

**Comment by enygren** [comment]
One approach is to treat the Alt-Svc record the same an AliasMode entry, either defined as such or if an Alt-Svc parameter is present.  This gets the behavior Eric describes.  This is good as it reduces duplication between Alt-Svc and HTTPS RRs.   A question would be how to handle conflicts between the two, specifically around alpn and port.  Do the Alt-Svc alpn and port just get ignored in this case, or are they used to filter the HTTPS RRs?

**Comment by bemasc** [comment]
Given that the draft is past DNSOP WGLC, I would prefer to avoid substantial revisions to the normative behavior.  I think we can settle this for now with an improved description of the intended behavior, with a clear path to revisions based on operational experience.

**Comment by martinthomson** [comment]
So I think that I like @enygren's suggestion to treat AltSvc as AliasMode, but I also agree with @bemasc about the unfortunate nature of the timing (and the potential for this to never finish if we start pulling on these strings).

> I would quibble with this. Servers can safely publish both; it's clients that make the choice to use ECH (and thus lose support for Alt-Svc). Making ECH secure without impairing Alt-Svc for non-ECH clients was a key design goal.

So servers don't make the decision alone: but they will make a decision based on what their clients do.  And there's a really bad outcome here where neither can be used if important clients make different decisions.

In any case, I'd really appreciate some text.  Maybe a) explain why you can't use AltSvc when you have an ECH config and b) make a soft promise that some future work will improve the situation.

**Comment by bemasc** [comment]
Please review proposed text at #327.

**Comment by ericorth** [comment]
That text still isn't convincing to my perspective that it would be a bad idea to query A/AAAA/HTTPS when we are making a connection attempt based on Alt-Svc, or that it would be worth the massive complexity to create such different DNS query behaviors.  And if HTTPS records on the Alt-Svc given host/port give us an ECH config, no reason not to use it.

If we have a "remote Alt-Svc" (that is one with a different host/port) for h3 (note that Chrome ignores Alt-Svc for anything but h3), we really only have two choices, whether to try to connect to that first, or to ignore it for a while as we first start a connection attempt against the original host/port and back out if we don't find an h3 endpoint in DNS.  Ignoring that Alt-Svc seems super silly to me because it's info that's been given to us via a fairly secure mechanism (unlike DoH, it's secure all the way to the server), it's the most likely fastest route to an h3 connection compared to the original host/port that we don't know whether or not it has h3 available until, and unless we skip the HTTPS queries when attempting the connection (which still seems silly to me), ECH can still work just fine if the domain publishes an HTTPS record at that domain, so all these arguments that we can't do this because it breaks ECH are just wrong.

(Worth noting, that if we have h3 Alt-Svc without a different host/port, we'll just make one h3 connection attempt that first tries to find h3 endpoints in HTTPS records, and if not, falls back to attempting h3 against the A/AAAA records, per traditional Alt-Svc behavior.  That seems consistent with both my current plans and the relevant rules in the current draft.)

How about "Clients that implement support for ECH MUST perform the HTTPS RR query first, and MUST only make use of Alt-Svc when operating in SVCB-optional mode (see {{ech-client-behavior}})" -> "... unless the client queries HTTPS and supports ECH when querying Alt-Svc names."

**Comment by ericorth** [comment]
I suppose one could argue that the current draft would even require us to attempt h2/http1.1 connections on HTTPS records before considering Alt-Svc.  That sounds even worse.  All the h3 logic is too tied up into a separate connection path.  We'd have to make an h3 connection attempt that backs out on not finding any h3-supporting HTTPS records.  Then a non-h3 connection attempt that backs out on not finding HTTPS records.  Then a connection attempt using the Alt-Svc info to attempt h3 against the A/AAAA records (after probably checking HTTPS records again).  (And then if the Alt-Svc-backed connection attempt fails or takes too long, another connection attempt to go back to non-h3 and continue even without HTTPS records.)

**Comment by enygren** [comment]
I think @ericorth has a good point that the current draft is dangerously underspecified.
The text:

> This specification does not alter the DNS records used when connecting 
> to an Alt-Svc hostname (typically A and/or AAAA only).

is vague enough to too much open to interpretation which risks interoperability challenges.
If we are seeing different clients interpreting it differently now, it may be better clarifying
now and confirming with the HTTP WG on desired behavior.  (This is in some ways
an IETF Last-Call topic as the issues are interactions with a spec from HTTP WG.)

Open questions we seem to land now seem to be:
- Can Alt-Svc be used at all with HTTPS RRs?  The desired answer seems to be yes for SVCB-optional at least.
- Should we do an HTTPS RR lookup for the name included in the Alt-Svc entry?  The desired answer seems to be yes (effectively looking at the Alt-Svc as an AliasMode record)
- What happens with the port in the Alt-Svc entry for the DNS lookup?  Does it impact the name being resolved?  I'd propose that it should be ignored for the purpose of the HTTPS RR name to lookup.
- What happens with the port in the Alt-Svc entry?  Either: 1) any port in the HTTPS RR should take precedence, but the one from the Alt-Svc gets used if none exists;  2) any port in the HTTPS RR takes precedence and the Alt-Svc one is always ignored even if the HTTPS RR lacks one; 3) the Alt-Svc record port is used as a filter
- What happens with the alpn in the Alt-Svc entry?  This is similar to the port with the same options (precedence; alt-svc alpn always ignored; acts as a filter)
- Does this draft need to "Update" the Alt-Svc RFC?  Answer depends on the above.

For port and ALPN, acting as a filter might have the cleanest semantics but I can see arguments both ways.

Other issues around Alt-Svc would make sense to defer to an Alt-Svc update, unless there's an import one that I missed.  @ericorth and @martinthomson , do these cover your areas of ambiguity?  Do you think we need to answer the above here and now?

**Comment by bemasc** [comment]
As a process matter, I don't think we should Update Alt-Svc, and should avoid any normative language that triggers such an Update.

As a technical matter, I don't think we can simply say that Alt-Svc triggers an HTTPS query, for several reasons:
* Security: Alt-Svc values are delivered securely, but the HTTPS RR is not.  Replacing a received Alt-Svc value with data from an HTTPS query could enable a network intermediary to override secure instructions in the Alt-Svc value.
* Targeting: Alt-Svc values are targeted to a single client, but the HTTPS RR is universal.  Overriding it would undo all the Alt-Svc targeting.
* Control/consistency: The Alt-Svc specification does not demand that the origin have full control or ownership of the Alt-host's DNS zone.  In some cases, the appropriate SvcParams for the origin and the Alt-host zones may be different and irreconcilable, especially if the Alt-host name is also used as its own origin or is shared by various origins with different requirements.

I also think it's counterintuitive: the Alt-Svc field value is logically parallel to a ServiceMode record, so there is no need for further service resolution.

These issues are likely surmountable, but solving them will require imposing new limits on how Alt-Svc can be configured, and what kinds of parameters are allowed to be defined in Alt-Svc, or perhaps defining some new Alt-Svc parameters.  I don't think we should try to figure out, and get consensus on, those modifications in this draft.

> I suppose one could argue that the current draft would even require us to attempt h2/http1.1 connections on HTTPS records before considering Alt-Svc.

I don't see any requirement for that ... except if ECH is in use, in which case Alt-Svc can't be considered at all.  In general, the priority ordering of connection attempts has little or no normative force.  We can certainly clarify if there is misleading text on this point.

**Comment by ericorth** [comment]
A lot of this likely comes down to intricacies that are unique to Chrome's processes and our only-partial support for Alt-Svc.  For us, Alt-Svc is just a pointer in triggering connection processes.  We use Alt-Svc to initiate an h3 connection attempt against a given host/port.  It's not really a sensible thing to "treat it like" any kind of DNS record because it is controlling different things at a different part of the process.  Any concerns around dealing with conflicting information between Alt-Svc and DNS are not a real concern for us because we do not use Alt-Svc that way and probably never will.  Any attempt to fully define how Alt-Svc and DNS are supposed to work together is probably going to be way too over-specified to meet our specific scenario.

Can we come up with a simple "MAY" addition to let Chrome do what it needs to do here without trying to come up with some complicated scheme to try to create fully-specified algorithm around it?

Or I suppose the alternative is that we argue that Chrome's Alt-Svc behavior is a unique not-fully-standard thing, and thus not actually making "use of Alt-Svc", and thus these HTTPS restrictions don't actually apply, and then I just implement what I need to implement and argue back if anybody ever tries to open a bug saying Chrome shouldn't be doing it.

> Can Alt-Svc be used at all with HTTPS RRs? The desired answer seems to be yes for SVCB-optional at least.
> Should we do an HTTPS RR lookup for the name included in the Alt-Svc entry? The desired answer seems to be yes (effectively looking at the Alt-Svc as an AliasMode record)

At least for Chrome's usage of Alt-Svc, I see no reason not to do this.  All Alt-Svc currently is is a pointer to some other host/port for an h3 connection attempt.  Completely different part of the process from DNS, so that's not really looking at it as an AliasMode record, but that might be the closest analogy when looking at the end result of Chrome behavior.

> What happens with the port in the Alt-Svc entry for the DNS lookup? Does it impact the name being resolved? I'd propose that it should be ignored for the purpose of the HTTPS RR name to lookup.

If Alt-Svc gave us a port, that is now part of the request.  Natural behavior is to use it the same as if the original request contained a port in the URL.  That is that if the port is not 443, initial HTTPS queries are made with a prefix to specify the port.  E.g., if Alt-Svc told us to attempt a connection to example.com:661, we then do a normal connection attempt for that host/port, including asking DNS if there's any HTTPS information for example.com:661.  Anything else would take a bunch of special-case complications to allow a connection attempt to use a port for everything except DNS.  This feels like a case that the HTTPS spec already has reasonable rules to deal with.

> What happens with the port in the Alt-Svc entry? Either: 1) any port in the HTTPS RR should take precedence, but the one from the Alt-Svc gets used if none exists; 2) any port in the HTTPS RR takes precedence and the Alt-Svc one is always ignored even if the HTTPS RR lacks one; 3) the Alt-Svc record port is used as a filter

Same thing as if a user typed in a URL with a port or if DNS aliases us to a name with a port.  We already have rules and algorithms to deal with this and don't need anything special.  If an HTTPS record at that port then tells us to connect to a different port, that's what we'll do.

> What happens with the alpn in the Alt-Svc entry? This is similar to the port with the same options (precedence; alt-svc alpn always ignored; acts as a filter)

With or without Alt-Svc, Chrome already has to do this filtering.  Due to complexities around connection sharing, h3 and non-h3 are separate connection flows with separate calls to DNS (but note that due to local caching, we'll generally only make the actual DNS queries once).  If we're in an h3 connection job, whether or not there was an Alt-Svc at the start of the process, we need to look through the results and only consider h3-supporting endpoints.  So while such filtering seems like an extra complication, for Chrome's needs, it's actually the simpler natural behavior.

> Does this draft need to "Update" the Alt-Svc RFC? Answer depends on the above.

Haven't read the Alt-Svc RFC recently enough to know if anything here completely violates its rules, but my guess would be that no such update is needed.  We're just talking about how DNS is handled as part of a connection attempt after using Alt-Svc to decide to make an h3 connection against some host/port.  Different not-actually-conflicting parts of the overall process.

> > I suppose one could argue that the current draft would even require us to attempt h2/http1.1 connections on HTTPS records before considering Alt-Svc.
> 
> I don't see any requirement for that ... except if ECH is in use, in which case Alt-Svc can't be considered at all. In general, the priority ordering of connection attempts has little or no normative force. We can certainly clarify if there is misleading text on this point.

We do not get DNS and know what names will have ECH configs until after we have started connection jobs.  We simply do not know if ECH will be in use until after we have set quite a bit of connection priority ordering.

I think there's an assumption that clients can always just do all the DNS and then can look at that together with Alt-Svc info and decide what to do.  Seems simple and reasonable, but when you consider all the complications of socket/connection pooling (especially with QUIC), that simply isn't how it works.

**Comment by enygren** [comment]
Ideally we'd define a behavior that clients can implement reasonably and in a well-defined manner.
If a major client like Chrome implements something different that Alt-Svc in response to the Alt-Svc header,
that breaks the ability of servers/operators to deploy Alt-Svc in a standard compliant manner without
also worrying about how Chrome uses that header but does something differently.

If www.example.com has an Alt-Svc which specifies svc.example.com:661 then:

* I hope/assume that Chrome still treats the origin host as "www.example.com" and validates certificates against www.example.com  (as anything other than this would badly break anyone trying to deploy Alt-Svc as specified).
* For HTTPS RR lookups, does it make sense to lookup "svc.example.com" or "_661._https.svc.example.com" ?   We need to pick one and specify either way or things become dangerously underspecified and hard to deploy.

As specified currently for HTTPS, it is always the origin that is used for HTTPS RR lookups, so the current specification doesn't really allow Alt-Svc to lookup an HTTPS RR for "svc.example.com" or "_661._https.svc.example.com" in the above case.

On the topic of "filtering" by ALPN, it sounds like this is effectively approach Chrome already takes?
ie, if an Alt-Svc is for "h3" then HTTPS RRs for "http/1.1" and "h2" will be ignored?
If an Alt-Svc is for "h2" then will HTTPS RRs for "h3" also be ignored?

**Comment by ericorth** [comment]
> I hope/assume that Chrome still treats the origin host as "www.example.com" and validates certificates against www.example.com (as anything other than this would badly break anyone trying to deploy Alt-Svc as specified).

Correct.

> For HTTPS RR lookups, does it make sense to lookup "svc.example.com" or "_661._https.svc.example.com" ? We need to pick one and specify either way or things become dangerously underspecified and hard to deploy.

Good point.  To me "_661._https.svc.example.com" is the obvious choice, and what I assumed was intended from the draft until this thread pointed out this situation to me.  "https://svc.example.com:661" is the full name for which you are looking up DNS information, and "_661._https.svc.example.com" is how you translate that to an HTTPS query.

> if an Alt-Svc is for "h3" then HTTPS RRs for "http/1.1" and "h2" will be ignored?

For the logic handling the h3 connection flow, yes, because that logic is already locked in and only capable of moving on to h3 connections after DNS.  So unless we completely rearchitect all Chrome connection logic (not a practical option), the only way to handle HTTPS info coming from the DNS stack is for the h3 connection logic to look through the results and pick the h3-supporting endpoints.

(But note that this connection flow might be running in parallel with non-h3 connection flows, typically with a headstart to the h3 flow to make that the preference.  The non-h3 flows are similarly already locked in and only capable of moving on to h1/h2 connections after DNS.)

> If an Alt-Svc is for "h2" then will HTTPS RRs for "h3" also be ignored?

In theory, that is what we'd likely have to do if we supported h2 Alt-Svc.  (But Chrome ignores Alt-Svc unless h3.)

**Comment by bemasc** [comment]
If we allow an HTTPS query for the Alt-host, that automatically imposes the following requirements:
* Clients that make such queries cannot rely on any security properties of securely delivered Alt-Svc information, since it can be overridden by a network intermediary.
* Servers that use both must not rely on client-targeting of any parameters other than the Alt-host, since they will be overridden by the untargeted SvcParams.
* Whoever writes the origin's HTTPS record must also control the Alt-host zone contents, to avoid conflicting SvcParams.  This could be difficult to support in current CDNs.

The first two considerations may not apply to Chrome's partial implementation, but for this draft we cannot ignore any of them.

I think HTTPS RRs and Alt-Svc overlap enough in their capabilities that doing both at once may not be worth the complexity and overhead.  I would prefer to avoid creating a lot of complexity in this draft trying to spell out how to integrate them, especially if operational experience may tell us that the wisest course is simply to disable Alt-Svc when HTTPS RRs are in use.

An Alt-Svc-bis could be specified with better alignment to HTTPS RRs, perhaps avoiding these issues.

**Comment by ericorth** [comment]
Keep researching/thinking this through, and I do not think we can make it work without exploding implementation complexity if we strictly comply with the current draft.  We simply cannot make all HTTPS queries before considering Alt-Svc unless we really overcomplicate the behavior around connection jobs, and it leads to silly stuff like having to attempt non-h3 before h3 Alt-Svc, just in case there are HTTPS records with ECH on the non-h3 endpoints when there wasn't anything usable for h3 (seems like a silly scenario to worry about).

And if we have Alt-Svc data, that's a strong enough signal that that host/port supports h3, that it'll only lead to performance issues to run through a whole bunch of other connection logic before attempting to consider the Alt-Svc, so we really need to be able to try it first, and we therefore need a mechanism to get ECH for those endpoints.

If we make HTTPS usable with Alt-Svc, that solves all these problems because Alt-Svc-based connections can reasonably get ECH or other HTTPS-only stuff from HTTPS records before any potential updates to Alt-Svc or additions to client support of Alt-Svc (which let's face it, might be an even bigger barrier than Alt-Svc updates), and therefore we don't need to restrict using Alt-Svc.

My proposal:

- HTTPS-supporting clients SHOULD (MAY?) query Alt-Svc host/portss alongside the A/AAAA queries when attempting to use Alt-Svc info (and query the entire Alt-Svc host including the port if different from default port).  MAY be skipped if all supported information was already read from the Alt-Svc.
- Clients querying HTTPS for an Alt-Svc host/ports different from the origin host/port MUST only consider endpoints compatible with any information the client supports using from Alt-Svc and that were read from the specific Alt-Svc info, other than the host/port name where redirects and such are still followed because DNS naturally owns the "where to go" question.  (Or maybe we say port must be compatible too and only host is special because host is the one thing that doesn't affect anything except DNS and is already mostly overridable through CNAME.)  In the Chrome case, this compatible endpoints stuff just means only considering h3 endpoints, as our implementation is naturally going to do anyway.  But I think this would also be sufficient for clients making more use of Alt-Svc including Alt-Svc features added in the future.  E.g., if some clients start accepting ECH keys read from Alt-Svc, if such an entry were provided in the Alt-Svc, the client would only be able to use HTTPS endpoints with a matching ECH config.
- Domains/servers/whatever that support ECH and have Alt-Svc with different hosts/ports from the origin SHOULD publish an HTTPS record on the Alt-Svc host/port with the ECH config.

I think that solves all the concerns and can be added to the draft without too much complexity.

> - Clients that make such queries cannot rely on any security properties of securely delivered Alt-Svc information, since it can be overridden by a network intermediary.

They can rely on it if we only allow compatible HTTPS rather than overriding HTTPS.

> - Servers that use both must not rely on client-targeting of any parameters other than the Alt-host, since they will be overridden by the untargeted SvcParams.

Again, not a problem for compatible merging instead of override merging.

> - Whoever writes the origin's HTTPS record must also control the Alt-host zone contents, to avoid conflicting SvcParams. This could be difficult to support in current CDNs.

If we go with compatible merging, I think the worst-case here is that there are no compatible HTTPS records, and thus it is no worse off than your proposal where HTTPS is not even queried.

**Comment by ericorth** [comment]
And thinking further through the implications of my proposal, as long as we add the last point (servers publishing HTTPS records with ECH config for the Alt-Svc host) into the current draft, I think it would then be fairly reasonable to leave the rest specifying all the override vs compatible etc behavior for a separate RFC, and Chrome would just be querying HTTPS as pre-standardized behavior in the meantime.  Goes a long way towards the goal of not changing things any more than absolutely necessary now in this draft.

**Comment by martinthomson** [comment]
> Clients querying HTTPS for an Alt-Svc host/ports different from the origin host/port MUST only consider endpoints compatible with any information the client supports using from Alt-Svc and that were read from the specific Alt-Svc info, other than the host/port name where redirects and such are still followed because DNS naturally owns the "where to go" question. 

I don't think that this is necessary.  We need to treat Alt-Svc and DNS both in the same way: as hints.

Yes, the information provided with Alt-Svc is authenticated, so it is arguably more trustworthy than what the DNS says (DNSSEC disclaimer, etc...).  However, I don't think that it helps.  It makes the system more complicated and brittle though.  Let's say that you learn that an ALPN of only "http/1.1" is supported for this alternative.  What then if the DNS says that "h2" is an option?  Maybe HTTP/2 was just added.  Does the client attempt HTTP/2 anyway?  How does it treat evidence that there was a downgrade?  Of course, ALPN is robust against that particular downgrade, but follow the logic with h3 instead of h2.

If you plot out the downgrade protection mechanisms you need here, I think you will find that this works better if you treat information about ALPN and protocol versions as informational.  No matter whether they come from Alt-Svc or DNS.  If the client needs to filter at each stage, you might find that nothing connects.  But if the hints are treated as just that, then the only things that matter are the final HTTPS/SVCB record (which determines what is attempted) and the authenticated connection (which is the ultimate source of authentic information that the client can use to work out if it was attacked).

> Domains/servers/whatever that support ECH and have Alt-Svc with different hosts/ports from the origin SHOULD publish an HTTPS record on the Alt-Svc host/port with the ECH config.

I think that is the right answer; it just means modeling Alt-Svc as equivalent to AliasMode.  So the same rationale for parallel A/AAAA and HTTPS queries applies as when Alt-Svc isn't in the mix.

**Comment by MikeBishop** [comment]
I think there's a fundamental flaw with treating Alt-Svc as an alias mode record, which is that the server name provided in Alt-Svc could have resources published under its own origin.  Alt-Svc provides an alternative network endpoint to contact the _original_ origin, and doesn't tell you anything about origins with the new hostname in the URL; an HTTPS lookup of the new hostname is talking about origins with that hostname.  Now, hopefully they overlap on things like ECH keys and supported protocols, but I'm not confident they are guaranteed to overlap.

Perhaps the right integration would be:

- Do an HTTPS lookup for the origin you intend to connect to (not the alternative).
- If you learn anything about the alternative host (such as ECH keys or additional supported protocols) from the HTTPS lookup, you can use that information.
- Prefer the host(s) indicated in the Alt-Svc, assuming they meet your criteria for use.  Specifically, if you require ECH for your connections and did not find ECH keys for a host, don't use it.

That allows servers to use Alt-Svc to indicate preferred alternatives to clients, but leaves the HTTPS path intact for providing information about the alternatives.

**Comment by bemasc** [comment]
> * If you learn anything about the alternative host (such as ECH keys or additional supported protocols) from the HTTPS lookup, you can use that information.

This seems like it creates a really weird interaction between Alt-Svc and DNS TTLs.

> if you require ECH for your connections

Yes, but should clients require ECH for Alt-Svc connections if they used ECH for the initial connection?

**Comment by enygren** [comment]
> I think there's a fundamental flaw with treating Alt-Svc as an alias mode record, which is that the server name provided in Alt-Svc could have resources published under its own origin. Alt-Svc provides an alternative network endpoint to contact the _original_ origin, and doesn't tell you anything about origins with the new hostname in the URL; an HTTPS lookup of the new hostname is talking about origins with that hostname. Now, hopefully they overlap on things like ECH keys and supported protocols, but I'm not confident they are guaranteed to overlap.
>
>     * Do an HTTPS lookup for the origin you intend to connect to (not the alternative).

This is an operational consideration.  The server configurations specified by both Alt-Svc alternative endpoints
and HTTPS records are decoupled from origin hosts.  There is no reason why a single alternative
endpoint couldn't handle quite a few different origins all sharing the same configuration.
(In fact, I could see this being common for some hosting providers with millions
of virtual sites to point the HTTPS records for all of those sites to the same HTTPS SvcMode 
record via a mixture of AliasMode and Alt-Svc records.)

>     * If you learn anything about the alternative host (such as ECH keys or additional supported protocols) from the HTTPS lookup, you can use that information.

You can't mix-and-match like this.  You really need to keep the alternative server name in the HTTPS record tied to the SvcParams in that exact same HTTPS record.

> Targeting: Alt-Svc values are targeted to a single client, but the HTTPS RR is universal. Overriding it would undo all the Alt-Svc targeting.

The targeting happens by returning different names.  If we treat the Alt-Svc similar to the target of an AliasMode record then the server can still use the client to decide which HTTPS SvcMode record it wants to name.  The difference is that the name may now point to an HTTPS record rather than just A/AAAA records.

**Comment by bemasc** [comment]
There are a ton of options here.
1. Current draft: No change to Alt-Svc.  Alt-Svc doesn't support ECH, so clients don't use Alt-Svc if ECH is on.
2. Alt-Svc as Alias-ish: Clients do an HTTPS query for the Alt-host.  Any ServiceMode record received overrides the Alt-Svc params.
3. Alt-Svc as Filter (@ericorth): Clients do an HTTPS query for the Alt-Host, but restrict any received ServiceMode records to be consistent with the Alt-Svc params before use (according to some criteria specified here).
4. Alt-host Overlap (@mikebishop).  Clients do an HTTPS query for the Origin, but only connect to the Alt-host.  If the Origin's SVCB chain crosses the Alt-host, you're in luck: go to Option 2.
5. New param: Add an Alt-Svc param of some sort to tell the client what to do (e.g. `ech=...`, `svcb-mode=...`).

A key consideration is whether server operators and origins who want working ECH (for clients who support it) have to worry about their Alt-Svc configuration.  AFAICT, only Option 1 allows server operators to enable ECH without thinking about Alt-Svc.  The other options make ECH insecure with some allowed Alt-Svc configurations.  In other words, they make ECH's security dependent not only on the origin zone, but also on the Alt-host zones.

I also think it's worth considering which of these options gives us the most room to adjust the behavior later, if we're not confident that we can find the right solution now.  My understanding is that only Option 1 lets us pursue all of the others in the future without compatibility risk.

**Comment by MikeBishop** [comment]
> A key consideration is whether server operators and origins who want working ECH (for clients who support it) have to worry about their Alt-Svc configuration. AFAICT, only Option 1 allows server operators to enable ECH without thinking about Alt-Svc. The other options make ECH insecure with some allowed Alt-Svc configurations.

Not sure I'd agree with that.  I think the starting point for any of these is that a client configured for ECH MUST disregard alternatives (regardless of provenance) for which it doesn't have ECH keys, provided there exists any endpoint for which it does possess ECH keys.  If it has ECH keys for the origin but hasn't found them for an alt-host, however we decide it should do so, then it should ignore its Alt-Svc entry and go with the contents of the HTTPS record.

If there's no way to combine info from one to the other, that means Alt-Svc always gets ignored if the origin publishes any options in HTTPS records with ECH keys.  I think that's an acceptable outcome for now.  It would be nice to define (in an updated Alt-Svc draft) how data from HTTPS records can be consumed while using an Alt-Svc entry to connect.

**Comment by martinthomson** [comment]
@MikeBishop, think of it as a chain or maybe a ratchet.  Where you enter the ratchet matters, because that is the name you will be using (in SNI and to authenticate the server).  Using AliasMode means that you can enter at any point (CNAME too, just to be complete).  

If you have A->B->C, then looking for B will have you connect to C and send B as an SNI.  Looking for A means sending an SNI of A to C instead.  If C has its own resources then you might enter at the chain and send an SNI of C to C.

As @enygren says, this chaining is likely to be a common configuration in practice (at least for small numbers of indirections).  

ECH is funny here only in that what C has to say determines whether you expect the alternative service D to have an ECH record.  If C has ECH, then you won't talk to D unless D (or anything it refers to) has ECH.  Alt-Svc is funny in that it doesn't really change the entrypoint.  So if you started with A and you are talking to C, then Alt-Svc to D (which might refer to E) still means that you are using A as the name.  And as you are using the same name, you should use the same ECH posture.  So even though you have been A->B->C->D->E, E still needs to provide ECH and answer correctly when addressed as A.

Seems fairly straightforward to me.  Of course, That doesn't mean changing the draft under discussion at all.

**Comment by bemasc** [comment]
I've expanded the proposal in #327 based on the feedback.  While it hasn't substantively changed, the new description relaxes some of the language, with the result that it resembles @MikeBishop's proposal.  (When strict ECH is enabled, clients can't escape the Service Bindings, but they can use Alt-Svc to steer within those bindings.)  I think the new text is also more amenable to updates that define how to do ECH with Alt-Svc.

**Comment by bemasc** [comment]
Apologies, I forgot to push my updated text.  #327 is now updated.

The key line in this proposed text (similar to the current draft) is "Origins can publish HTTPS records without any need to modify their existing Alt-Svc arrangements.".  In my view, there is essentially one alternative, which is to say "Clients MAY use HTTPS records from the Alt-Svc hostname, so servers SHOULD ensure that those records are consistent with the origin's".  We cannot delay this decision: once clients and servers are deployed on either assumption, it is locked in.

The second option makes it unsafe for multiple origins to share an Alt-Svc hostname if their HTTPS records might differ (even in the future, due to long Alt-Svc cache lifetimes).  If the CDN allows customers to have different HTTPS records, the customers must be partitioned into separate Alt-Svc hostname groups.  If the CDN allows customers to publish their own HTTPS records, it must only use Alt-Svc hostnames that are under the control of the customer.  This seems like it could become a significant burden for CDN operators and customers.

This is our choice: do we require client implementors to let the origin's HTTPS record influence Alt-Svc behaviors (current draft), or do we require server operators to publish matching HTTPS records on the Alt-Svc hostname?  Perhaps someone with more CDN experience can comment on how difficult this would be.

**Comment by enygren** [comment]
> If the CDN allows customers to publish their own HTTPS records, it must only use Alt-Svc 
> hostnames that are under the control of the customer. 

I don't follow this?  Why doesn't the Alt-Svc hostname just want to be under the same
domain of control as the HTTPS SvcMode record?  ie, a CDN would return an 
Alt-Svc hostname that it selects and has an HTTPS RR.  In a multi-CDN case, 
CDN Alpha would return one Alt-Svc hostname corresponding to its own HTTPS RRs, and CDN Bravo would return one Alt-Svc hostname corresponding to its own HTTPS RRs, and an origin not on a CDN would do the same.

I guess the case which could be tricky when Alt-Svc is to ":443" so uses the origin name as the Alt-Svc hostname so this could switch CDNs, but this can already switch CDN's with Alt-Svc today in this case.  Worst case seems to be that the client has to abandon the Alt-Svc and go back to HTTPS RRs targetted by the origin hostname.

**Comment by bemasc** [comment]
> Why doesn't the Alt-Svc hostname just want to be under the same domain of control as the HTTPS SvcMode record?

Yes, that's what I mean.  In the alternative designs, if a hosting provider permits its customers to publish their own ServiceMode records, it must only use Alt-Svc hostnames that they control.  In some CDN arrangements (e.g. static IP hosting), the norm is currently for all records to live in the customer's zone.

If the CDN publishes its own ServiceMode records (as expected in the common case), then this consideration does not apply.  However, even if the CDN controls all the records, the alternative designs still create a synchronization requirement if the CDN can generate different ServiceMode records for different customers.

**Comment by ericorth** [comment]
Summarizing the options from the perspective of client behavior, I see two options here:

1. Use Alt-Svc to get the request host/port and pass that off to the connection logic similarly to a non-Alt-Svc host/port (which is the origin), and as a result of treating it similarly to non-Alt-Svc stuff, the connection/DNS logic makes the normal A/AAAA/HTTPS queries  (all for the request host/port which is the alt-svc hostname) and ECH is supported via ECH keys in the resultingly queried HTTPS records.  From an external-behavior perspective, this works out as "treat Alt-Svc as an alias".

   Pros:
   - Minimizes client complexity because it minimizes differences between Alt-Svc and non-Alt-Svc connections.  And minimizing client complexity means more adoption and less accidental adoption of spec-non-compliant behavior.  (Somebody is bound to query HTTPS for Alt-Svc hostnames whether or not that is the intention, so it's safer if we define how that is done.)
   - Easier partial/transition adoption of HTTPS.  E.g., if some client just wants to query HTTPS to get ECH keys (either because that is all the client cares about or because that is just the first step towards full support of all HTTPS-allowed behavior), that client can mostly just add HTTPS queries alongside A/AAAA queries, and add the logic to handle HTTPS results including using ECH.  No need to alter bigger-picture parts of the connection logic that a client likely would not need to alter if not (yet) supporting things like DNS-based upgrade to h3.
   - If we think we're eventually going to want HTTPS queries with Alt-Svc, it's safer to define now how that works.  Vs if we say for now that it's forbidden, servers may make assumptions that clients will never ever do it, e.g. by specifying HTTPS records that are incompatible if the client got to the host/port through an Alt-Svc.  We wouldn't be able to safely change the interaction without a new Alt-Svc or a new HTTPS.
   
   Cons:
   - Complicates/slows standardization of this spec.  Need to coordinate with the HTTPS WG because, depending on the specifics of the Alt-Svc spec that I haven't studied in too much detail lately, this either updates Alt-Svc, or in the least, reduces some flexibility around potential future changes to Alt-Svc (because any future Alt-Svc parameters need to account for how they may interact with information from HTTPS).
   - As Ben noted above, could force servers to ensure unique Alt-Svc hostnames per origin to allow different HTTPS information on each of those hostnames.  But I would assume this isn't actually a problem unless some implementor (especially a CDN) has specific problems with it.  It's just the general nature of HTTPS records that if you have origin-specific parameters, that you shouldn't let any chains from other origins point to the domain with those HTTPS records.  Seems the same issues to me as CNAME/HTTPS alias redirection.  If somebody wants origin-specific parameters, they also need to update their CNAMEs to point to origin-specific domains rather than a big shared domain.

2. Alter a bunch of connection logic to block querying HTTPS for Alt-Svc names and ensure Alt-Svc is not used if the origin has HTTPS records (or at least HTTPS records that would result in using ECH).

   Pros:
   - Other than maybe cleaning up some explanations, this is pretty much what we currently have written in the draft, so it's the best path to getting the spec standardized as soon as possible.
   - No coordination/changes needed with Alt-Svc.  Alt-Svc and HTTPS are effectively treated as two completely isolated mechanisms with interactions disallowed.

   Cons:
   - Except for extremely simple clients that just do all the DNS for the origin, and then attempt to connect to the results, this option requires a lot of non-obvious complex changes to client implementations and all their connection logic.  Real-world client implementations, especially browsers, have a lot of complexity in the connection logic, and you really do not want to mess with that any more than necessary.  I previously said this may require insurmountable complexity around kicking off a whole bunch of semi-concurrent connection jobs for all the different combinations of parameters that may get used for the eventual connection (with/without Alt-Svc, h3 vs non-h3, etc).  We think we may have found a way to make it more in the realm of possible by altering some of our h3 jobs to run concurrent DNS requests for Alt-Svc name and the origin and decide on cancelling out of using Alt-Svc names depending on DNS results from the origin, but there's still a lot of hidden complexity here that I would prefer not to introduce to already-complex code, and I might easily later discover some reason this plan doesn't work.

**Comment by ericorth** [comment]
> > Clients querying HTTPS for an Alt-Svc host/ports different from the origin host/port MUST only consider endpoints compatible with any information the client supports using from Alt-Svc and that were read from the specific Alt-Svc info, other than the host/port name where redirects and such are still followed because DNS naturally owns the "where to go" question.
> 
> I don't think that this is necessary. We need to treat Alt-Svc and DNS both in the same way: as hints.
> 
> Yes, the information provided with Alt-Svc is authenticated, so it is arguably more trustworthy than what the DNS says (DNSSEC disclaimer, etc...). However, I don't think that it helps. It makes the system more complicated and brittle though. Let's say that you learn that an ALPN of only "http/1.1" is supported for this alternative. What then if the DNS says that "h2" is an option? Maybe HTTP/2 was just added. Does the client attempt HTTP/2 anyway? How does it treat evidence that there was a downgrade? Of course, ALPN is robust against that particular downgrade, but follow the logic with h3 instead of h2.

Servers shouldn't send the client an HTTPS record saying "you can use h3 if you go here" if they don't want the client connecting h3 to that endpoint.  So from that perspective, this seems reasonable to me.  Once that record has been received by the client, after following some chain from the origin, it seems reasonable to use it, even if one of the chain links said "go over to this hostname if you want h2".

But if you get to an h3 endpoint via an h2 Alt-Svc, is it also acceptable for the client to attempt an h2 connection to that endpoint, as would seem reasonable if we treat these purely as hints, or would h2 only be usable on falling back to the Alt-Svc A/AAAA per SVCB-optional behavior (which would seem essentially equivalent to the "override" behavior previously discussed)? Unless we justify it as the Alt-Svc ALPNs getting merged with any HTTPS ALPNs, attempting h2 would seem to be changing the "Clients SHOULD NOT attempt connection to a service endpoint whose SVCB ALPN set does not contain any supported protocols" rule.

And what about any Alt-Svc params added in the future where it really doesn't make sense to treat them as hints or allow overriding? I suppose it could be argued that this is safe to decide in the future when/if such params are ever added (but only if the HTTPS WG is okay with this implicit requirement being placed on future development of Alt-Svc).

**Comment by bemasc** [comment]
I still don't see why implementing support for #326 is hard.  HTTPS queries are for the origin, and Alt-Svc doesn't change the origin.  With Alt-Svc, every connection job has a "target" and an origin (which are the same except in remote alt-svc), and each job should query A/AAAA for the former and HTTPS for the latter, whether they're the same or different.

However, I've written up #329 as an alternative to #326, describing what I think @ericorth is proposing.  Hopefully that lets us compare the options head-to-head.

**Comment by bemasc** [comment]
I believe this is resolved by #329.
